Add `Scan` input/output type annotations to `debugprint` output

brandonwillard opened the pull request:
Closes #945

Here are the results of printing a somewhat complicated `Scan` under these changes:
```python

Sum{acc_dtype=float64} [id A]
 |for{cpu,scan_fn} [id B] (outer_out_nit_sot-0)
 |Elemwise{scalar_minimum,no_inplace} [id C] (outer_in_nit_sot-0)
 | |Subtensor{int64} [id D]
 | | |Shape [id E]
 | | | |Subtensor{int64::} [id F] 'coefficients[0:]'
 | | |   |coefficients [id G]
 | | |   |ScalarConstant{0} [id H]
 | | |ScalarConstant{0} [id I]
 | |Subtensor{int64} [id J]
 |   |Shape [id K]
 |   | |Subtensor{int64::} [id L]
 |   |   |ARange{dtype='int64'} [id M]
 |   |   | |TensorConstant{0} [id N]
 |   |   | |TensorConstant{10} [id O]
 |   |   | |TensorConstant{1} [id P]
 |   |   |ScalarConstant{0} [id Q]
 |   |ScalarConstant{0} [id R]
 |Subtensor{:int64:} [id S] (outer_in_seqs-0)
 | |Subtensor{int64::} [id F] 'coefficients[0:]'
 | |ScalarFromTensor [id T]
 |   |Elemwise{scalar_minimum,no_inplace} [id C]
 |Subtensor{:int64:} [id U] (outer_in_seqs-1)
 | |Subtensor{int64::} [id L]
 | |ScalarFromTensor [id V]
 |   |Elemwise{scalar_minimum,no_inplace} [id C]
 |Elemwise{scalar_minimum,no_inplace} [id C] (outer_in_nit_sot-0)
 |A [id W] (outer_in_non_seqs-0)
 |k [id X] (outer_in_non_seqs-1)
 
 Inner graphs:
 
 for{cpu,scan_fn} [id B] (outer_out_nit_sot-0)
 >Elemwise{mul,no_inplace} [id Y] (inner_out_nit_sot-0)
 > |InplaceDimShuffle{x} [id Z]
 > | |coefficients[t] [id BA] -> [id S] (inner_in_seqs-0)
 > |Elemwise{pow,no_inplace} [id BB]
 >   |Subtensor{int64} [id BC]
 >   | |Subtensor{int64::} [id BD]
 >   | | |for{cpu,scan_fn} [id BE] (outer_out_sit_sot-0)
 >   | | | |k_copy [id BF] -> [id X] (inner_in_non_seqs-1) (n_steps)
 >   | | | |IncSubtensor{Set;:int64:} [id BG] (outer_in_sit_sot-0)
 >   | | | | |AllocEmpty{dtype='float64'} [id BH]
 >   | | | | | |Elemwise{add,no_inplace} [id BI]
 >   | | | | | | |k_copy [id BF] -> [id X] (inner_in_non_seqs-1)
 >   | | | | | | |Subtensor{int64} [id BJ]
 >   | | | | | |   |Shape [id BK]
 >   | | | | | |   | |Rebroadcast{(0, False)} [id BL]
 >   | | | | | |   |   |InplaceDimShuffle{x,0} [id BM]
 >   | | | | | |   |     |Elemwise{second,no_inplace} [id BN]
 >   | | | | | |   |       |A_copy [id BO] -> [id W] (inner_in_non_seqs-0)
 >   | | | | | |   |       |InplaceDimShuffle{x} [id BP]
 >   | | | | | |   |         |TensorConstant{1.0} [id BQ]
 >   | | | | | |   |ScalarConstant{0} [id BR]
 >   | | | | | |Subtensor{int64} [id BS]
 >   | | | | |   |Shape [id BT]
 >   | | | | |   | |Rebroadcast{(0, False)} [id BL]
 >   | | | | |   |ScalarConstant{1} [id BU]
 >   | | | | |Rebroadcast{(0, False)} [id BL]
 >   | | | | |ScalarFromTensor [id BV]
 >   | | | |   |Subtensor{int64} [id BJ]
 >   | | | |A_copy [id BO] -> [id W] (inner_in_non_seqs-0) (outer_in_non_seqs-0)
 >   | | |ScalarConstant{1} [id BW]
 >   | |ScalarConstant{-1} [id BX]
 >   |InplaceDimShuffle{x} [id BY]
 >     |<TensorType(int64, ())> [id BZ] -> [id U] (inner_in_seqs-1)
 
 for{cpu,scan_fn} [id BE] (outer_out_sit_sot-0)
 >Elemwise{mul,no_inplace} [id CA] (inner_out_sit_sot-0)
 > |<TensorType(float64, (None,))> [id CB] -> [id BG] (inner_in_sit_sot-0)
 > |A_copy [id CC] -> [id BO] (inner_in_non_seqs-0)
```

- [x] Decouple/remove the `Scan`-specific code from `aesara.printing`
  - We could add a dispatch function&mdash;based on the `Op`&mdash;that computes the `extra_information` used by this new feature.

ricardoV94 reviewed (dismissed):
Looks good, glad to get rid of those empty strings at the end

brandonwillard commented:
> Looks good, glad to get rid of those empty strings at the end

Yes, that was too annoying to leave in place.

Anyway, I just renamed some of the new variables/functions and added an option to enable/disable the extra information; otherwise, I think this is ready to go.

My main goal with this change is to make it easier for people to understand how `Scan`s are organized, because I believe that will help remove a lot of the mystery/confusion surrounding this complicated `Op`.  

Plus, I find myself creating `ScanArgs` objects simply to decode which inputs/outputs are assigned to which inner-IO patterns (e.g. mit-mots, sit-sots, nit-nots, etc.), and these changes simplify that process dramatically.
